pennylane_lightning_gpu/src/simulator/StateVectorCudaMPI.hpp

```diff
@@ -941,6 +955,364 @@ class StateVectorCudaMPI
+            for (size_t j = 0; j < matrix.size(); j++) {
+                nnzs.push_back(matrix[j].values.size());
+            }
+        }
+
+        size_t local_nnz = mpi_manager_.scatter<size_t>(nnzs, 0)[0];
+
+        CSRMatrix<Precision, index_type> localCSRMatrix(local_num_rows,
+                                                        local_nnz);
+
+        if (mpi_manager_.getRank() == root) {
+            localCSRMatrix.values = matrix[0].values;
+            localCSRMatrix.csrOffsets = matrix[0].csrOffsets;
+            localCSRMatrix.columns = matrix[0].columns;
+        }
+
+        for (size_t k = 1; k < num_col_blocks; k++) {
+            size_t dest = k;
+            size_t source = 0;
+
+            if (mpi_manager_.getRank() == 0 && matrix[k].values.size()) {
+                mpi_manager_.Send<std::complex<Precision>>(matrix[k].values,
+                                                           dest);
+            } else if (mpi_manager_.getRank() == k && local_nnz) {
+                mpi_manager_.Recv<std::complex<Precision>>(
+                    localCSRMatrix.values, source);
+            }
+
+            if (mpi_manager_.getRank() == 0 && matrix[k].values.size()) {
+                mpi_manager_.Send<index_type>(matrix[k].csrOffsets, dest);
+            } else if (mpi_manager_.getRank() == k && local_nnz) {
+                mpi_manager_.Recv<index_type>(localCSRMatrix.csrOffsets,
+                                              source);
+            }
+
+            if (mpi_manager_.getRank() == 0 && matrix[k].values.size()) {
+                mpi_manager_.Send<index_type>(matrix[k].columns, dest);
+            } else if (mpi_manager_.getRank() == k && local_nnz) {
+                mpi_manager_.Recv<index_type>(localCSRMatrix.columns, source);
+            }
+        }
+        return localCSRMatrix;
+    }
+
+    /**
+     * @brief Convert a global CSR (Compress Sparse Row) format matrix into
+     * local blocks. This operation should be conducted on the rank 0.
+     *
+     * @tparam index_type Integer type used as indices of the sparse matrix.
+     * @param num_row_blocks Number of local blocks per global row.
+     * @param num_col_blocks Number of local blocks per global column.
+     * @param csrOffsets_ptr Pointer to the array of row offsets of the sparse
+     * matrix. Array of size csrOffsets_size.
+     * @param columns_ptr Pointer to the array of column indices of the sparse
+     * matrix. Array of size numNNZ
+     * @param values_ptr Pointer to the array of the non-zero elements
+     *
+     * @return auto A vector of vector of CSRMatrix.
+     */
+    template <class index_type>
+    auto splitCSRMatrix(const size_t &num_row_blocks,
+                        const size_t &num_col_blocks,
+                        const index_type *csrOffsets_ptr,
+                        const index_type *columns_ptr,
+                        const std::complex<Precision> *values_ptr)
```

Comment:
Do we want to explicitly access the raw pointers here? Do we need this for GPU device access, or CPU side? If CPU, we can probably avoid raw pointers (use a vector/span/etc). If GPU, we can probably use the DataBuffer class instead.

Reply:
These data are on the host side. I put it in pointer because the data are pointer from the pybind layer and no additional operations are conducted on these pointers. If needed, I can update it.